src/sumbuddy/__main__.py

```diff
@@ -2,33 +2,123 @@
 import os
 import hashlib
 import csv
+import fnmatch
 from tqdm import tqdm
 
 def md5_checksum(file_path):
+    """
+    Calculate the MD5 checksum of a file
+    """
     hash_md5 = hashlib.md5()
     with open(file_path, "rb") as f:
         for chunk in iter(lambda: f.read(4096), b""):
             hash_md5.update(chunk)
     return hash_md5.hexdigest()
 
-def get_checksums(input_directory, output_filepath):
+def read_patterns(filepath):
+    """
+    Read patterns from an include or ignore file, one pattern per line
+    """
+    with open(filepath, 'r') as f:
+        patterns = [line.strip() for line in f if line.strip()]
+    return patterns
+
+
+def should_include(filepath, patterns, root, include_mode=False):
+    """
+    Determine if a file should be included based on patterns and the root directory.
+    Excludes hidden files and directories by default unless explicitly included.
+    """
+    relative_filepath = os.path.relpath(filepath, start=root)
+
+    # Default behavior when no patterns are passed: exclude all hidden files and directories
+    if not patterns:
+        return not relative_filepath.split(os.sep)[0].startswith('.')
+
+    # Handle include mode
+    if include_mode:
+        # Include files that match any pattern
+        return any(fnmatch.fnmatch(relative_filepath, pattern) for pattern in patterns)
+
+    # Handle exclude mode with negation
+    for pattern in patterns:
+        if pattern.startswith('!'):
+            if fnmatch.fnmatch(relative_filepath, pattern[1:]):
+                # Negation pattern matches, include this file
+                return True  
+
+    # Handle exclude mode
+    for pattern in patterns:
+        if fnmatch.fnmatch(relative_filepath, pattern):
+            # Pattern matches for ignore, exclude this file
+            return False  
+
+    return not relative_filepath.split(os.sep)[0].startswith('.')
+
+
+def gather_file_paths(input_directory, output_filepath, ignore_file=None, include_file=None):
```

Comment:
I feel like there are too many functions in this file and we should move some into other helper files so `__main__` isn't so overloaded. Perhaps just keep `get_checksums` and `main` here.